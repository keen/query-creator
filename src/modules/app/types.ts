import {
  APP_START,
  STORE_SCHEMAS,
  FETCH_PROJECT_DETAILS,
  FETCH_PROJECT_DETAILS_SUCCESS,
  FETCH_PROJECT_DETAILS_ERROR,
} from './constants';

export interface AppStartAction {
  type: typeof APP_START;
}

export interface StoreSchemasAction {
  type: typeof STORE_SCHEMAS;
}

export interface FetchProjectDetailsAction {
  type: typeof FETCH_PROJECT_DETAILS;
}

export interface FetchProjectDetailsSuccessAction {
  type: typeof FETCH_PROJECT_DETAILS_SUCCESS;
}

export interface FetchProjectDetailsErrorAction {
  type: typeof FETCH_PROJECT_DETAILS_ERROR;
  payload: {
    error: Error;
  };
}

export type AppActions =
  | AppStartAction
  | StoreSchemasAction
  | FetchProjectDetailsAction
  | FetchProjectDetailsSuccessAction
  | FetchProjectDetailsErrorAction;
