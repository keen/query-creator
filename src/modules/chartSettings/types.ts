import { UPDATE_CHART_SETTINGS, RESET_CHART_SETTINGS } from './constants';

export type ChartSettings = {
  stepLabels: string[];
};

export type ReducerState = ChartSettings;

export interface UpdateChartSettingsAction {
  type: typeof UPDATE_CHART_SETTINGS;
  payload: Partial<ChartSettings>;
}

export interface ResetChartSettingsAction {
  type: typeof RESET_CHART_SETTINGS;
}

export type ChartSettingsActions =
  | UpdateChartSettingsAction
  | ResetChartSettingsAction;
