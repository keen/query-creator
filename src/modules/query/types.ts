import {
  SERIALIZE_QUERY,
  SET_QUERY,
  SELECT_EVENT_COLLECTION,
  SELECT_ANALYSIS,
  SELECT_TARGET_PROPERTY,
  SET_PERCENTILE,
  SET_TIMEFRAME,
  SET_GROUP_BY,
  SET_ORDER_BY,
  SET_INTERVAL,
  SET_LIMIT,
  SET_EXTRACTION_LIMIT,
  SET_PROPERTY_NAMES,
  SET_FILTERS,
  SELECT_TIMEZONE,
  ADD_FUNNEL_STEP,
  SET_FUNNEL_STEPS,
  SET_FUNNEL_STEP_FILTERS,
  CLONE_FUNNEL_STEP,
  REMOVE_FUNNEL_STEP,
  UPDATE_FUNNEL_STEP,
  SELECT_FUNNEL_STEP_EVENT_COLLECTION,
  CHANGE_FUNNEL_STEPS_ORDER,
  ADD_FUNNEL_STEP_FILTER,
  UPDATE_FUNNEL_STEP_FILTER,
  REMOVE_FUNNEL_STEP_FILTER,
  UPDATE_FUNNEL_STEP_TIMEZONE,
  RESET_EXTRACTION,
  RESET_QUERY,
  ADD_FILTER,
  UPDATE_FILTER,
  REMOVE_FILTER,
} from './constants';

import {
  Timezones,
  Timeframe,
  OrderBy,
  ExtractionProperty,
  FunnelStep,
  Filter,
} from '../../types';
import { Analysis } from '../../types';

export type ReducerState = {
  eventCollection?: string;
  targetProperty?: string;
  percentile?: number;
  timezone?: number | Timezones;
  groupBy?: string | string[];
  orderBy?: OrderBy[];
  limit?: number;
  timeframe?: Timeframe;
  interval?: string;
  analysisType: Analysis;
  filters?: Filter[];
  steps?: FunnelStep[];
  propertyNames?: ExtractionProperty[];
  latest?: number;
};

export type InitialQuery = Omit<ReducerState, 'propertyNames'> & {
  propertyNames?: string | string[];
};

export interface SetQueryAction {
  type: typeof SET_QUERY;
  payload: {
    query: Partial<ReducerState>;
  };
}

export interface SerializeQueryAction {
  type: typeof SERIALIZE_QUERY;
  payload: {
    query: Partial<InitialQuery>;
  };
}

export interface ResetQueryAction {
  type: typeof RESET_QUERY;
}

export interface AddFilterAction {
  type: typeof ADD_FILTER;
  payload: { id: string };
}

export interface RemoveFilterAction {
  type: typeof REMOVE_FILTER;
  payload: { id: string };
}

export interface UpdateFilterAction {
  type: typeof UPDATE_FILTER;
  payload: { id: string; filter: Partial<Filter> };
}

export interface SetExtractionLimitAction {
  type: typeof SET_EXTRACTION_LIMIT;
  payload: {
    limit: number;
  };
}

export interface ResetExtractionAction {
  type: typeof RESET_EXTRACTION;
}

export interface SelectEventCollectionAction {
  type: typeof SELECT_EVENT_COLLECTION;
  payload: {
    name: string;
  };
}

export interface SelectAnalysisAction {
  type: typeof SELECT_ANALYSIS;
  payload: {
    analysis: Analysis;
  };
}

export interface SelectTargetPropertyAction {
  type: typeof SELECT_TARGET_PROPERTY;
  payload: {
    property: string;
  };
}

export interface SelectTimezoneAction {
  type: typeof SELECT_TIMEZONE;
  payload: {
    timezone: Timezones;
  };
}

export interface SetPercentileAction {
  type: typeof SET_PERCENTILE;
  payload: {
    percentile: number;
  };
}

export interface SetPropertyNamesAction {
  type: typeof SET_PROPERTY_NAMES;
  payload: {
    propertyNames: ExtractionProperty[];
  };
}

export interface SetTimeframeAction {
  type: typeof SET_TIMEFRAME;
  payload: {
    timeframe: Timeframe;
  };
}

export interface SetGroupByAction {
  type: typeof SET_GROUP_BY;
  payload: {
    groupBy: string | string[] | undefined;
  };
}

export interface SetIntervalAction {
  type: typeof SET_INTERVAL;
  payload: {
    interval: string | undefined;
  };
}

export interface SetOrderByAction {
  type: typeof SET_ORDER_BY;
  payload: {
    orderBy: OrderBy[] | undefined;
  };
}

export interface SetLimitAction {
  type: typeof SET_LIMIT;
  payload: {
    limit: number | undefined;
  };
}

export interface SetFiltersAction {
  type: typeof SET_FILTERS;
  payload: {
    filters: Filter[];
  };
}

export interface AddFunnelStepAction {
  type: typeof ADD_FUNNEL_STEP;
  payload: {
    id: string;
  };
}

export interface CloneFunnelStepAction {
  type: typeof CLONE_FUNNEL_STEP;
  payload: {
    cloneId: string;
    newId: string;
  };
}

export interface SelectFunnelStepEventCollectionAction {
  type: typeof SELECT_FUNNEL_STEP_EVENT_COLLECTION;
  payload: { name: string };
}

export interface RemoveFunnelStepAction {
  type: typeof REMOVE_FUNNEL_STEP;
  payload: {
    stepId: string;
  };
}

export interface UpdateFunnelStepAction {
  type: typeof UPDATE_FUNNEL_STEP;
  payload: {
    stepId: string;
    properties: Partial<FunnelStep>;
  };
}

export interface ChangeFunnelStepsOrderAction {
  type: typeof CHANGE_FUNNEL_STEPS_ORDER;
  payload: { steps: FunnelStep[] };
}

export interface SetFunnelSteps {
  type: typeof SET_FUNNEL_STEPS;
  payload: { steps: FunnelStep[] };
}

export interface SetFunnelStepFilters {
  type: typeof SET_FUNNEL_STEP_FILTERS;
  payload: { stepId: string; filters: Filter[] };
}

export interface AddFunnelStepFilterAction {
  type: typeof ADD_FUNNEL_STEP_FILTER;
  payload: {
    stepId: string;
    filterId: string;
  };
}

export interface UpdateFunnelStepFilterAction {
  type: typeof UPDATE_FUNNEL_STEP_FILTER;
  payload: {
    stepId: string;
    filterId: string;
    properties: Filter;
  };
}

export interface RemoveFunnelStepFilterAction {
  type: typeof REMOVE_FUNNEL_STEP_FILTER;
  payload: {
    stepId: string;
    filterId: string;
  };
}

export interface UpdateFunnelStepTimezoneAction {
  type: typeof UPDATE_FUNNEL_STEP_TIMEZONE;
  payload: {
    stepId: string;
    timezone: Timezones;
  };
}

export type QueryActions =
  | SerializeQueryAction
  | SetQueryAction
  | ResetQueryAction
  | SelectEventCollectionAction
  | SelectAnalysisAction
  | SelectTargetPropertyAction
  | SetPercentileAction
  | SetGroupByAction
  | SetOrderByAction
  | SetIntervalAction
  | SetLimitAction
  | SetExtractionLimitAction
  | SetPropertyNamesAction
  | SetTimeframeAction
  | SetFiltersAction
  | SelectTimezoneAction
  | SetFunnelSteps
  | SetFunnelStepFilters
  | AddFunnelStepAction
  | UpdateFunnelStepAction
  | RemoveFunnelStepAction
  | SelectFunnelStepEventCollectionAction
  | ChangeFunnelStepsOrderAction
  | CloneFunnelStepAction
  | AddFunnelStepFilterAction
  | UpdateFunnelStepFilterAction
  | RemoveFunnelStepFilterAction
  | UpdateFunnelStepTimezoneAction
  | ResetExtractionAction
  | AddFilterAction
  | RemoveFilterAction
  | UpdateFilterAction;
