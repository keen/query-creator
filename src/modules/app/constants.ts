export const APP_START = '@app/APP_START';
export const STORE_SCHEMAS = '@app/STORE_SCHEMAS';
export const FETCH_PROJECT_DETAILS = '@app/FETCH_PROJECT_DETAILS';
export const FETCH_PROJECT_DETAILS_SUCCESS =
  '@app/FETCH_PROJECT_DETAILS_SUCCESS';
export const FETCH_PROJECT_DETAILS_ERROR = '@app/FETCH_PROJECT_DETAILS_ERROR';
