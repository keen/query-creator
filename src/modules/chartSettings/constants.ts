export const UPDATE_CHART_SETTINGS = '@query-creator/UPDATE_CHART_SETTINGS';
export const RESET_CHART_SETTINGS = '@query-creator/RESET_CHART_SETTINGS';
