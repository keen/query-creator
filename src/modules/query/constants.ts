import { FunnelStep } from '../../types';

export const SET_QUERY = '@query/SET_QUERY';
export const SERIALIZE_QUERY = '@query/SERIALIZE_QUERY';
export const POST_PROCESSING_FINISHED = '@query/POST_PROCESSING_FINISHED';
export const RESET_QUERY = '@query/RESET_QUERY';
export const SELECT_EVENT_COLLECTION = '@query/SELECT_EVENT_COLLECTION';
export const SELECT_ANALYSIS = '@query/SELECT_ANALYSIS';
export const SELECT_TARGET_PROPERTY = '@query/SELECT_TARGET_PROPERTY';
export const SET_PERCENTILE = '@query/SET_PERCENTILE';
export const SET_TIMEFRAME = '@query/SET_TIMEFRAME';
export const SET_GROUP_BY = '@query/SET_GROUP_BY';
export const SET_INTERVAL = '@query/SET_INTERVAL';
export const SET_ORDER_BY = '@query/SET_ORDER_BY';
export const SET_LIMIT = '@query/SET_LIMIT';
export const RESET_EXTRACTION = '@query/RESET_EXTRACTION';
export const SET_EXTRACTION_LIMIT = '@query/SET_EXTRACTION_LIMIT';
export const SET_PROPERTY_NAMES = '@query/SET_PROPERTY_NAMES';
export const SET_FILTERS = '@query/SET_FILTERS';
export const SELECT_TIMEZONE = '@query/SELECT_TIMEZONE';
export const ADD_FUNNEL_STEP = '@query/ADD_FUNNEL_STEP';
export const SELECT_FUNNEL_STEP_EVENT_COLLECTION =
  '@query/SELECT_FUNNEL_STEP_EVENT_COLLECTION';
export const REMOVE_FUNNEL_STEP = '@query/REMOVE_FUNNEL_STEP';
export const UPDATE_FUNNEL_STEP = '@query/UPDATE_FUNNEL_STEP';
export const CLONE_FUNNEL_STEP = '@query/CLONE_FUNNEL_STEP';
export const SET_FUNNEL_STEPS = '@query/SET_FUNNEL_STEPS';
export const SET_FUNNEL_STEP_FILTERS = '@query/SET_FUNNEL_STEP_FILTERS';
export const CHANGE_FUNNEL_STEPS_ORDER = '@query/CHANGE_FUNNEL_STEP_ORDER';
export const ADD_FUNNEL_STEP_FILTER = '@query/ADD_FUNNEL_STEP_FILTER';
export const UPDATE_FUNNEL_STEP_FILTER = '@query/UPDATE_FUNNEL_STEP_FILTER';
export const REMOVE_FUNNEL_STEP_FILTER = '@query/REMOVE_FUNNEL_STEP_FILTER';
export const UPDATE_FUNNEL_STEP_TIMEZONE = '@query/UPDATE_FUNNEL_STEP_TIMEZONE';
export const ADD_FILTER = '@query/ADD_FILTER';
export const UPDATE_FILTER = '@query/UPDATE_FILTER';
export const REMOVE_FILTER = '@query/REMOVE_FILTER';

export const DEFAULT_ANALYSIS = 'count';
export const DEFAULT_TIMEFRAME = 'this_14_days';
export const DEFAULT_TIMEZONE = 'UTC';
export const DEFAULT_EXTRACTION_LIMIT = 100;

export const DEFAULT_FILTER = {
  propertyName: undefined,
  operator: undefined,
  propertyValue: undefined,
  propertyType: undefined,
};

export const DEFAULT_FUNNEL_STEP: FunnelStep = {
  eventCollection: undefined,
  actorProperty: undefined,
  timeframe: DEFAULT_TIMEFRAME,
  timezone: DEFAULT_TIMEZONE,
  filters: [],
  inverted: false,
  optional: false,
  withActors: false,
};
