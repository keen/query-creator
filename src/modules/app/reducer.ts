import { ReducerState, AppActions } from './types';

import { SET_QUERY_READINESS } from './constants';

export const initialState: ReducerState = {
  isQueryReady: false,
};

export const appReducer = (
  state: ReducerState = initialState,
  action: AppActions
) => {
  switch (action.type) {
    case SET_QUERY_READINESS:
      return {
        ...state,
        isQueryReady: action.payload.isQueryReady,
      };
    default:
      return state;
  }
};
