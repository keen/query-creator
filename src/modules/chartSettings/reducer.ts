import { ReducerState, ChartSettingsActions } from './types';

import { UPDATE_CHART_SETTINGS, RESET_CHART_SETTINGS } from './constants';

export const initialState: ReducerState = {
  stepLabels: [],
};

export const chartSettingsReducer = (
  state: ReducerState = initialState,
  action: ChartSettingsActions
) => {
  switch (action.type) {
    case UPDATE_CHART_SETTINGS:
      return {
        ...state,
        ...action.payload,
      };
    case RESET_CHART_SETTINGS:
      return initialState;
    default:
      return state;
  }
};
