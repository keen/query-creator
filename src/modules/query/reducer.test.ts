import { queryReducer, initialState } from './reducer';

import {
  resetExtraction,
  setQuery,
  setLimit,
  setTimeframe,
  setExtractionLimit,
  setPropertyNames,
  updateFunnelStep,
  removeFunnelStep,
  addFunnelStep,
  cloneFunnelStep,
  addFunnelStepFilter,
  changeFunnelStepsOrder,
  updateFunnelStepFilter,
  updateFunnelStepTimezone,
  removeFunnelStepFilter,
  setGroupBy,
  setOrderBy,
  setInterval,
  setFilters,
  selectAnalysis,
  selectTargetProperty,
  selectEventCollection,
  setPercentile,
} from './actions';
import { DEFAULT_FUNNEL_STEP } from './constants';

import { ReducerState } from './types';
import { OrderBy, Filter } from '../../types';

test('reset extraction properties to initial state', () => {
  const action = resetExtraction();
  const state = queryReducer(
    {
      ...initialState,
      propertyNames: [{ id: 'id', propertyName: 'country' }],
    },
    action
  );

  expect(state).toEqual(initialState);
});

test('set query properties', () => {
  const properties = {
    analysisType: 'extraction',
    groupBy: ['country'],
  };
  const action = setQuery(properties as Partial<ReducerState>);
  const state = queryReducer(initialState, action);

  expect(state).toEqual({
    ...initialState,
    ...properties,
  });
});

test('set query limit', () => {
  const action = setLimit(100);
  const { limit } = queryReducer(initialState, action);

  expect(limit).toEqual(100);
});

test('set extraction limit', () => {
  const action = setExtractionLimit(45);
  const { latest } = queryReducer(initialState, action);

  expect(latest).toEqual(45);
});

test('set extraction property names', () => {
  const properties = [{ id: 'id', propertyName: 'surname' }];
  const action = setPropertyNames(properties);

  const state = queryReducer(initialState, action);

  expect(state.propertyNames).toEqual(properties);
});

test('set query timeframe', () => {
  const timeframe = 'this_100_days';
  const action = setTimeframe(timeframe);
  const state = queryReducer(initialState, action);

  expect(state.timeframe).toEqual(timeframe);
});

test('set query order by properties', () => {
  const orderBy: OrderBy[] = [
    {
      propertyName: 'result',
      direction: 'ASC',
    },
  ];

  const action = setOrderBy(orderBy);
  const state = queryReducer(initialState, action);

  expect(state.orderBy).toEqual(orderBy);
});

test('set query interval', () => {
  const interval = 'monthly';
  const action = setInterval(interval);
  const state = queryReducer(initialState, action);

  expect(state.interval).toEqual(interval);
});

test('set group by properties', () => {
  const groupBy = ['name', 'surname'];

  const action = setGroupBy(groupBy);
  const state = queryReducer(initialState, action);

  expect(state.groupBy).toEqual(groupBy);
});

test('set percentile', () => {
  const percentile = 85;

  const action = setPercentile(percentile);
  const state = queryReducer(initialState, action);

  expect(state.percentile).toEqual(percentile);
});

test('set analysis', () => {
  const analysis = 'funnel';

  const action = selectAnalysis(analysis);
  const state = queryReducer(initialState, action);

  expect(state.analysisType).toEqual(analysis);
});

test('set target property', () => {
  const property = 'keen.id';

  const action = selectTargetProperty(property);
  const state = queryReducer(initialState, action);

  expect(state.targetProperty).toEqual(property);
});

test('set event collection', () => {
  const collection = 'logins';

  const action = selectEventCollection(collection);
  const state = queryReducer(initialState, action);

  expect(state.eventCollection).toEqual(collection);
});

test('set filters', () => {
  const filters = [
    { propertyName: 'state', operator: 'ne', propertyValue: 'Arizona' },
  ] as Filter[];

  const action = setFilters(filters);
  const state = queryReducer(initialState, action);

  expect(state.filters).toEqual(filters);
});

test('update funnel step', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
        id: 'asd123',
      },
    ],
  };

  const action = updateFunnelStep('asd123', {
    withActors: true,
    eventCollection: 'logins',
  });

  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toMatchSnapshot();
});

test('removes funnel step', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
        id: 'asd123',
      },
    ],
  };

  const action = removeFunnelStep('asd123');
  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toEqual([]);
});

test('add funnel step', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
      },
    ],
  };

  const action = addFunnelStep('qwe123');
  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toMatchSnapshot();
});

test('inherits settings from previous funnel step', () => {
  const stepsState = {
    steps: [
      {
        id: 'step1',
        ...DEFAULT_FUNNEL_STEP,
        timeframe: 'previous_1_year',
        eventCollection: 'clicks',
      },
    ],
  };

  const action = addFunnelStep('step2');
  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toMatchSnapshot();
});

test('clone funnel step', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
        id: 'asd123',
      },
    ],
  };

  const action = cloneFunnelStep('asd123', 'qwe456');
  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toMatchSnapshot();
});

test('add funnel step filter', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
        id: 'asd123',
      },
    ],
  };

  const action = addFunnelStepFilter('asd123', 'qwe456');
  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toMatchSnapshot();
});

test('update funnel step filter', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
        id: 'asd123',
        filter: [
          {
            id: 'qwe456',
            propertyName: 'id',
            operator: 'eq',
            propertyValue: 'city',
          },
        ],
      },
    ],
  };

  const action = updateFunnelStepFilter('asd123', 'qwe456', {
    propertyName: 'id',
    operator: 'ne',
    propertyValue: 'name',
  });
  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toMatchSnapshot();
});

test('remove funnel step filter', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
        id: 'asd123',
        filter: [
          {
            id: 'qwe456',
            propertyName: 'id',
            operator: 'eq',
            propertyValue: 'city',
          },
        ],
      },
    ],
  };

  const action = removeFunnelStepFilter('asd123', 'qwe456');
  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toMatchSnapshot();
});

test('updates funnel steps order', () => {
  const funnelSteps = [
    {
      ...DEFAULT_FUNNEL_STEP,
      eventCollection: 'purchases',
      id: 'id2',
      inverted: true,
      optional: true,
    },
    {
      ...DEFAULT_FUNNEL_STEP,
      eventCollection: 'clicks',
      id: 'id1',
    },
  ];

  const action = changeFunnelStepsOrder(funnelSteps);
  const { steps } = queryReducer(initialState, action);

  const [firstStep] = steps;

  expect(firstStep.inverted).toBeFalsy();
  expect(firstStep.optional).toBeFalsy();
});

test('update funnel step timezone', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
        id: 'asd123',
      },
    ],
  };

  const action = updateFunnelStepTimezone('asd123', 'US/Central');
  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toMatchSnapshot();
});
