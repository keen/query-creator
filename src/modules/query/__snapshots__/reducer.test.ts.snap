// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add funnel step 1`] = `
Array [
  Object {
    "actorProperty": undefined,
    "eventCollection": "clicks",
    "filters": Array [],
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": false,
  },
  Object {
    "actorProperty": undefined,
    "eventCollection": undefined,
    "filters": Array [],
    "id": "qwe123",
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": false,
  },
]
`;

exports[`add funnel step filter 1`] = `
Array [
  Object {
    "actorProperty": undefined,
    "eventCollection": "clicks",
    "filters": Array [
      Object {
        "id": "qwe456",
        "operator": undefined,
        "propertyName": undefined,
        "propertyType": undefined,
        "propertyValue": undefined,
      },
    ],
    "id": "asd123",
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": false,
  },
]
`;

exports[`clone funnel step 1`] = `
Array [
  Object {
    "actorProperty": undefined,
    "eventCollection": "clicks",
    "filters": Array [],
    "id": "asd123",
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": false,
  },
  Object {
    "actorProperty": undefined,
    "eventCollection": "clicks",
    "filters": Array [],
    "id": "qwe456",
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": false,
  },
]
`;

exports[`inherits settings from previous funnel step 1`] = `
Array [
  Object {
    "actorProperty": undefined,
    "eventCollection": "clicks",
    "filters": Array [],
    "id": "step1",
    "inverted": false,
    "optional": false,
    "timeframe": "previous_1_year",
    "timezone": undefined,
    "withActors": false,
  },
  Object {
    "actorProperty": undefined,
    "eventCollection": undefined,
    "filters": Array [],
    "id": "step2",
    "inverted": false,
    "optional": false,
    "timeframe": "previous_1_year",
    "timezone": undefined,
    "withActors": false,
  },
]
`;

exports[`remove funnel step filter 1`] = `
Array [
  Object {
    "actorProperty": undefined,
    "eventCollection": "clicks",
    "filter": Array [
      Object {
        "id": "qwe456",
        "operator": "eq",
        "propertyName": "id",
        "propertyValue": "city",
      },
    ],
    "filters": Array [],
    "id": "asd123",
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": false,
  },
]
`;

exports[`update funnel step 1`] = `
Array [
  Object {
    "actorProperty": undefined,
    "eventCollection": "logins",
    "filters": Array [],
    "id": "asd123",
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": true,
  },
]
`;

exports[`update funnel step filter 1`] = `
Array [
  Object {
    "actorProperty": undefined,
    "eventCollection": "clicks",
    "filter": Array [
      Object {
        "id": "qwe456",
        "operator": "eq",
        "propertyName": "id",
        "propertyValue": "city",
      },
    ],
    "filters": Array [],
    "id": "asd123",
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": false,
  },
]
`;

exports[`update funnel step timezone 1`] = `
Array [
  Object {
    "actorProperty": undefined,
    "eventCollection": "clicks",
    "filters": Array [],
    "id": "asd123",
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": "US/Central",
    "withActors": false,
  },
]
`;
